### DISCUSSION TITLE
StrongSwan Android - How to debug native code

### DISCUSSION THREAD

**Discussion by adepaoli** [strongswan/strongswan#956] [Q&A]
I'm not able to debug the native code (ie charonservice.c or strongswan/src/libcharon/sa/ikev2/authenticators/pubkey_authenticator.c)
What I should do?

**Comment by tobiasbrunner** [comment]
It's probably over ten years ago since I last debugged the native parts of the Android app, so I can't really help you, sorry. If you modified something that can be debugged on regular Linux (e.g. the pubkey authenticator), try that. If not, maybe add some log statements?

**Reply by adepaoli** [reply]
I need to understand and debug how the authenticator works. Any suggestion?
We have some issues with a FortiGate IKEv2 and CERT client authentication.

**Reply by tobiasbrunner** [reply]
And your first instinct is to actually throw a debugger at this? Start by reading the logs and looking at the certificates more closely.

**Reply by adepaoli** [reply]
Dear Tobias, we have a "special" use case to solve. 
The ServerCertificate cannot be for each "FortiGate" so we have a single certificate for every fortigate and 1 single certificate for each tablet.
When we generate the ServerCertificate we use --san "xxx.xxx.xxx" that could not be the same IP of each fortigate.
I'm investigating how i could bypass the validation of the SAN IP during the authentication of the client.
Any idea?
[charon_ip192.168.10.5_cersan192.168.10.6.log](https://github.com/strongswan/strongswan/files/8266393/charon_ip192.168.10.5_cersan192.168.10.6.log)

**Reply by adepaoli** [reply]
BTW: I figure out how to debug the Native Code using the latest Android Studio, but still working to find how to bypass the validation...

**Reply by tobiasbrunner** [reply]
> The ServerCertificate cannot be for each "FortiGate"

What does that mean?

> When we generate the ServerCertificate we use --san "xxx.xxx.xxx" that could not be the same IP of each fortigate.

Using FQDNs as SANs is usually a lot more flexible (unless the client explicitly uses an IP address as server address and remote identity). You can also put additional SANs in the certificates if that makes configuration on the client easier. The identity sent by the server still has to be contained in the certificate, though (see below).

> I'm investigating how i could bypass the validation of the SAN IP during the authentication of the client.

You are not actually at that point in the log. The server responded with its IP address as identity and a certificate is searched that contains it. However, it would seem that the certificate the server sent does not actually contain that IP as SAN. So either make sure the certificate does, or change the server's identity to one that's already contained in the certificate.

**Reply by adepaoli** [reply]
I end up to this code
	if (status == NOT_FOUND)
	{
		DBG1(DBG_IKE, "no trusted %N public key found for '%Y'",
			 key_type_names, key_type, id);
	}

When the server SAN IP is different from the Server Gateway IP.

**Reply by adepaoli** [reply]
How i can get the "string IP" from an ideintification_t of type ID_IPV4_ADDR?

**Reply by tobiasbrunner** [reply]
> When the server SAN IP is different from the Server Gateway IP.

And why can't you change that? Add the actual identity to the certificate or change the identity to one already in the certificate? Identities and SANs don't have to be IP addresses.

> How i can get the "string IP" from an ideintification_t of type ID_IPV4_ADDR?

I guess you could print it (`%Y`). But why?
